src/core/Ekubo.sol

```diff
@@ -0,0 +1,486 @@
+    }
+}
+
+abstract contract Ekubo is SettlerAbstract {
+    using UnsafeMath for uint256;
+    using UnsafeMath for int256;
+    using CreditDebt for int256;
+    using Ternary for bool;
+    using SafeTransferLib for IERC20;
+    using NotesLib for NotesLib.Note;
+    using NotesLib for NotesLib.Note[];
+    using StateLib for StateLib.State;
+    using UnsafeEkuboCore for IEkuboCore;
+
+    constructor() {
+        assert(BASIS == Encoder.BASIS);
+        assert(BASIS == Decoder.BASIS);
+        assert(ETH_ADDRESS == Decoder.ETH_ADDRESS);
+    }
+
+    //// How to generate `fills` for Ekubo
+    ////
+    //// Linearize your DAG of fills by doing a topological sort on the tokens involved. In the
+    //// topological sort of tokens, when there is a choice of the next token, break ties by
+    //// preferring a token if it is the lexicographically largest token that is bought among fills
+    //// with sell token equal to the previous token in the topological sort. Then sort the fills
+    //// belonging to each sell token by their buy token. This technique isn't *quite* optimal, but
+    //// it's pretty close. The buy token of the final fill is special-cased. It is the token that
+    //// will be transferred to `recipient` and have its slippage checked against `amountOutMin`. In
+    //// the event that you are encoding a series of fills with more than one output token, ensure
+    //// that at least one of the global buy token's fills is positioned appropriately.
+    ////
+    //// Take care to note that while Ekube represents the native asset of the chain as
+    //// the address of all zeroes, Settler represents this as the address of all `e`s. You must use
+    //// Settler's representation. The conversion is performed by Settler before making calls to Ekubo
+    ////
+    //// Now that you have a list of fills, encode each fill as follows.
+    //// First encode the `bps` for the fill as 2 bytes. Remember that this `bps` is relative to the
+    //// running balance at the moment that the fill is settled.
+    //// Second, encode the packing key for that fill as 1 byte. The packing key byte depends on the
+    //// tokens involved in the previous fill. The packing key for the first fill must be 1;
+    //// i.e. encode only the buy token for the first fill.
+    ////   0 -> sell and buy tokens remain unchanged from the previous fill (pure multiplex)
+    ////   1 -> sell token remains unchanged from the previous fill, buy token is encoded (diamond multiplex)
+    ////   2 -> sell token becomes the buy token from the previous fill, new buy token is encoded (multihop)
+    ////   3 -> both sell and buy token are encoded
+    //// Obviously, after encoding the packing key, you encode 0, 1, or 2 tokens (each as 20 bytes),
+    //// as appropriate.
+    //// The remaining fields of the fill are mandatory.
+    //// Third, encode the config of the pool as 32 bytes
+    //// Fourth, encode the skipAhead to use in the swap as 32 bytes
```

Comment:
Maybe explain the semantics of these fields by copying Ekubo's documentation? Or at least link to Ekubo's documentation

Reply:
Ekubo's doc is not that good to have a section explaining them but let me add comments